dali/operators/math/expressions/broadcasting_test.cc

```diff
@@ -0,0 +1,188 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include "dali/operators/math/expressions/broadcasting.h"
+#include "dali/core/tensor_shape.h"
+#include "dali/kernels/common/utils.h"
+
+namespace dali {
+namespace test {
+
+TEST(ArithmeticOpsBroadcastingTest, BroadcastShape) {
+  auto test = [](TensorShape<> expected, TensorShape<> a, TensorShape<> b) {
+    TensorShape<> result0, result1;
+    ASSERT_TRUE(CanBroadcast(a, b));
+    ASSERT_TRUE(CanBroadcast(b, b));
+    BroadcastShape(result0, a, b);
+    EXPECT_EQ(expected, result0);
+    BroadcastShape(result1, a, b);
+    EXPECT_EQ(expected, result1);
+  };
+  test({1, 3, 2}, {3, 2}, {1, 3, 1});
+  test({10, 10, 2}, {10, 1, 2}, {1, 10, 2});
+  test({10, 10, 2}, {10, 10, 2}, {});
+  test({10, 10, 3}, {10, 10, 1}, {3});
+  test({}, {}, {});
+  test({1}, {}, {1});
+
+  auto test_fail = [](TensorShape<> a, TensorShape<> b) {
+    TensorShape<> result;
+    ASSERT_FALSE(CanBroadcast(a, b));
+    ASSERT_FALSE(CanBroadcast(b, a));
+    ASSERT_THROW(BroadcastShape(result, a, b), std::runtime_error);
+    ASSERT_THROW(BroadcastShape(result, b, a), std::runtime_error);
+  };
+  test_fail({3, 2}, {2, 3});
+  test_fail({2}, {2, 3});
+}
+
+TEST(ArithmeticOpsBroadcastingTest, BroadcastTensorListShape) {
+  auto test = [](TensorListShape<> expected, TensorListShape<> a, TensorListShape<> b) {
+    TensorListShape result0, result1;
+    ASSERT_TRUE(CanBroadcast(a, b));
+    ASSERT_TRUE(CanBroadcast(b, a));
+    BroadcastShape(result0, a, b);
+    EXPECT_EQ(expected, result0);
+    BroadcastShape(result1, a, b);
+    EXPECT_EQ(expected, result1);
+  };
+  test({{2, 2, 3}}, {{2, 2, 3}}, {{2, 1, 3}});
+  test({{1, 3, 2}, {3, 2, 4}, {4, 2, 1}},
+       {{1, 3, 2}, {3, 2, 1}, {4, 1, 1}}, {{1, 1, 2}, {3, 2, 4}, {1, 2, 1}});
+  test({{1, 3, 2}, {3, 2, 4}, {2, 4, 1}},
+       {{1, 3, 2}, {3, 2, 1}, {2, 1, 1}}, {{1, 1, 2}, {3, 2, 4}, {1, 4, 1}});
```

Comment:
Probably it would be nice to see a test for TLS with different sample_dim of the inputs.